### PR TITLE
JS: support simple callbacks for setting up Express route handlers

### PR DESCRIPTION
**Pull request by erik-krogh**
Inspired by [this route handler](https://github.com/jakealbaugh/discogs-aggregator/blob/25efb4169f5082dd97b1e4e80869b42bf9b16fc5/app/server/components/Routes.js#L78). (Found using ATM)

I tried to see if I could [find more route-handlers using this addition](https://lgtm.com/query/8471273632731467076/).  
But so far I've not found anything other than the motivating-example. 

TODO: 
 - [x] Add test
 - [ ] Evaluation

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/Express.qll [hunk 1]
```diff
@@ -130,6 +130,12 @@ module Express {
         ) and
         t = t2
       )
+      or
+      exists(DataFlow::CallNode call, int i, DataFlow::FunctionNode forwarder |
+        result.(HTTP::RouteHandlerCandidate).flowsTo(call.getArgument(i)) and
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/Express.qll:135]:
```suggestion
      exists(DataFlow::CallNode call, DataFlow::FunctionNode forwarder |
        result.(HTTP::RouteHandlerCandidate) = call.getCallback(_)) and
```

**Comment by erik-krogh** [comment]
> I wonder if type-tracking of req/res instead be convinced to go through such wrapper functions?

It can, and that is probably a better approach. 

I've implemented it in `RequestSource::ref()` for now, but I feel it belongs somewhere else (although I'm unsure where). 

Doing the [same thing for all calls is a bad idea](https://lgtm.com/query/6332013482799928277/), as our callgraph might blow up in e.g. `angular.forEach()`.

**Comment by erik-krogh** [comment]
[An evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1594201537487) looks fine. 

I'll check out `--metric TaintSources.ql`, move the new type-tracking steps into a helper predicate, and then un-draft this PR.

**Comment by erik-krogh** [comment]
[A rather large evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1594728827746) with `--metric TaintSources.ql` was unable to find new taint-sources. 

I'll just close this PR for now because of the extremely limited impact. 
I'll take it up again if someone thinks the PR is worth landing.

**Pull request closed without merging**